chia/util/memory_profiler.py

```diff
@@ -0,0 +1,245 @@
+    log.info("Starting memory profiler. saving to %s" % profile_dir)
+    profile_dir.mkdir(parents=True, exist_ok=True)
+
+    try:
+        tracemalloc.start(30)
+
+        counter = 0
+
+        while True:
+            # this will throw CancelledError when we're exiting
+            await asyncio.sleep(60)
+            snapshot = tracemalloc.take_snapshot()
+            snapshot.dump(str(profile_dir / f"heap-{counter:05d}.profile"))
+            log.info(f"Heap usage: {tracemalloc.get_traced_memory()[0]/1000000:0.3f} MB profile {counter:05d}")
+            counter += 1
+    finally:
+        tracemalloc.stop()
+
+
+if __name__ == "__main__":
+    import sys
+    from functools import lru_cache
+    from subprocess import check_call
+    from sys import stdout
+
+    from colorama import Back, Fore, Style, init
+
+    g_next_id: int = 0
+
+    profile_dir = pathlib.Path(sys.argv[1])
+    init(strip=False)
+
+    def print_timeline() -> None:
+        counter = 0
+        try:
+            while True:
+                snapshot = tracemalloc.Snapshot.load(str(profile_dir / f"heap-{counter:05d}.profile"))
+                # the total memory usage in MB
+                total = sum(st.size for st in snapshot.statistics("filename")) / 1000000.0
+
+                if total > 150:
+                    color = Fore.RED + Style.BRIGHT
+                elif total > 120:
+                    color = Fore.MAGENTA + Style.BRIGHT
+                elif total > 90:
+                    color = Fore.YELLOW + Style.BRIGHT
+                elif total > 60:
+                    color = Style.BRIGHT
+                elif total < 30:
+                    color = Fore.GREEN
+                else:
+                    color = ""
+
+                quantized = int(total // 20)
+                print(f"{counter:05d}: {color}{total:3.0f} MB {Back.WHITE} {' ' * quantized}{Style.RESET_ALL}")
+
+                counter += 1
+        except Exception as e:
+            print(e)
+
+    def color(pct: float) -> str:
+        return f"{int((100.-pct)//10)+1}"
+
+    def fontcolor(pct: float) -> str:
+        if pct > 80 or pct < 20:
+            return "white"
+        else:
+            return "black"
+
+    @lru_cache(maxsize=10000)
+    def resolve_function(file: str, line: int) -> str:
+        try:
+            with open(file, "r") as f:
+                all_lines: List[str] = []
+                for row in f:
+                    all_lines.append(row)
+
+            # line numbers start at 1
+            while line > 0:
+                line -= 1
+                current = all_lines[line]
+                if " def " in current or current.startswith("def "):
+                    return current.split("def ")[1].split("(")[0]
+            return file.rsplit("/", 1)[1]
+        except Exception:
+            return "<unknown>"
+
+    def analyze_slot(slot: int) -> None:
+        file = str(profile_dir / f"heap-{slot:05d}.profile")
+        output_file = f"chia-heap-{slot}"
```

Comment:
```suggestion
        output_file = str(profile_dir / f"heap-{slot:05d}")
```

This lets the resulting files end up next to their input files in the profile dir.

Reply:
I don't have a strong opinion about this. my workflow of using these is that the resulting png file is temporary anyway and it's easy to open from the terminal's cwd. It's possible this workflow could be improved, and certainly when running A/B tests it's good to have labels (not just timestamps) of which branch was tested.

It's not obvious that the `~/.chia/mainnet/` folder is the best place to put the output of the analysis though. I would think it might be more appropriate in a directory under cwd.

Reply:
I think it would be reasonable to improve this workflow as part of, or after, the tools have been moved out of the production code

Reply:
> It's not obvious that the `~/.chia/mainnet/` folder is the best place to put the output of the analysis though.

Well, we also store the profiling data there currently so imo we should either move the profiling data to where the results are currently or the results to where the profiling data is currently?